From minchan.kim@gmail.com  Thu Jan 14 17:13:55 2010
From: Minchan Kim <minchan.kim@gmail.com>
Date: Mon, 11 Jan 2010 16:15:53 +0900
Subject: Staging: ramzswap: Free memory when create_device is failed
To: Greg Kroah-Hartman <greg@kroah.com>
Cc: Nitin Gupta <ngupta@vflare.org>, linux-mm <linux-mm@kvack.org>
Message-ID: <20100111161553.3acebae9.minchan.kim@barrios-desktop>


If create_device is failed, it can't free gendisk and request_queue of
preceding devices. It cause memory leak.

This patch fixes it.

Signed-off-by: Minchan Kim <minchan.kim@gmail.com>
CC: Nitin Gupta <ngupta@vflare.org>
Signed-off-by: Greg Kroah-Hartman <gregkh@suse.de>

---
 drivers/staging/ramzswap/ramzswap_drv.c |   21 +++++++++++++--------
 1 file changed, 13 insertions(+), 8 deletions(-)

--- a/drivers/staging/ramzswap/ramzswap_drv.c
+++ b/drivers/staging/ramzswap/ramzswap_drv.c
@@ -1292,7 +1292,7 @@ static struct block_device_operations ra
 	.owner = THIS_MODULE,
 };
 
-static void create_device(struct ramzswap *rzs, int device_id)
+static int create_device(struct ramzswap *rzs, int device_id)
 {
 	mutex_init(&rzs->lock);
 	INIT_LIST_HEAD(&rzs->backing_swap_extent_list);
@@ -1301,7 +1301,7 @@ static void create_device(struct ramzswa
 	if (!rzs->queue) {
 		pr_err("Error allocating disk queue for device %d\n",
 			device_id);
-		return;
+		return 0;
 	}
 
 	blk_queue_make_request(rzs->queue, ramzswap_make_request);
@@ -1313,7 +1313,7 @@ static void create_device(struct ramzswa
 		blk_cleanup_queue(rzs->queue);
 		pr_warning("Error allocating disk structure for device %d\n",
 			device_id);
-		return;
+		return 0;
 	}
 
 	rzs->disk->major = ramzswap_major;
@@ -1331,6 +1331,7 @@ static void create_device(struct ramzswa
 	add_disk(rzs->disk);
 
 	rzs->init_done = 0;
+	return 1;
 }
 
 static void destroy_device(struct ramzswap *rzs)
@@ -1368,16 +1369,20 @@ static int __init ramzswap_init(void)
 	/* Allocate the device array and initialize each one */
 	pr_info("Creating %u devices ...\n", num_devices);
 	devices = kzalloc(num_devices * sizeof(struct ramzswap), GFP_KERNEL);
-	if (!devices) {
-		ret = -ENOMEM;
+	if (!devices)
 		goto out;
-	}
 
 	for (i = 0; i < num_devices; i++)
-		create_device(&devices[i], i);
-
+		if (!create_device(&devices[i], i)) {
+			ret = i;
+			goto free_devices;
+		}
 	return 0;
+free_devices:
+	for (i = 0; i < ret; i++)
+		destroy_device(&devices[i]);
 out:
+	ret = -ENOMEM;
 	unregister_blkdev(ramzswap_major, "ramzswap");
 	return ret;
 }
